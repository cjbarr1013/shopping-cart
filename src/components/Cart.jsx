import PropTypes from 'prop-types';
import styles from '../styles/Cart.module.css';
import CartCard from './CartCard';
import CartSummary from './CartSummary';
import { useOutletContext } from 'react-router-dom';

function Cart() {
  const { cart, handleCartChange, subtotal } = useOutletContext();

  return (
    <div className={styles.cartContainer}>
      <div className={styles.cart}>
        <div className={styles.cardContainer}>
          {cart.length === 0 ?
            <p>Your cart is empty.</p>
          : cart.map((product) => (
              <CartCard
                key={product.id}
                product={product}
                handleCartChange={handleCartChange}
              ></CartCard>
            ))
          }
        </div>
        <CartSummary subtotal={subtotal}></CartSummary>
      </div>
    </div>
  );
}

Cart.propTypes = {
  cart: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      quantity: PropTypes.number.isRequired,
    })
  ).isRequired,
  subtotal: PropTypes.number.isRequired,
};

export default Cart;
