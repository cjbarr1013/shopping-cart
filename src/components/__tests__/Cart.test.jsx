import { vi, describe, it, expect, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Cart from '../Cart';

// Mock child components
vi.mock('../CartCard', () => ({
  default: ({ product }) => <div>{`Mock CartCard: ${product.name}`}</div>,
}));

vi.mock('../CartSummary', () => ({
  default: ({ subtotal }) => (
    <div>{`Mock CartSummary: Subtotal is $${subtotal}`}</div>
  ),
}));

// Define mockUseOutletContext before using it
const mockUseOutletContext = vi.fn();

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useOutletContext: () => mockUseOutletContext(),
  };
});

// Reset the mock before each test
beforeEach(() => {
  mockUseOutletContext.mockReset();
});

describe('Cart Component', () => {
  it('renders the empty cart message when the cart is empty', () => {
    // Return an empty cart for this test
    mockUseOutletContext.mockReturnValue({
      cart: [],
      handleCartChange: vi.fn(),
      subtotal: 0,
    });

    render(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>
    );

    expect(screen.getByText('Your cart is empty.')).toBeInTheDocument();
    expect(
      screen.getByRole('link', { name: /go back to store/i })
    ).toBeInTheDocument();
  });

  it('renders the "Go back to store" link correctly', () => {
    mockUseOutletContext.mockReturnValue({
      cart: [],
      handleCartChange: vi.fn(),
      subtotal: 0,
    });

    render(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>
    );

    const link = screen.getByRole('link', { name: /go back to store/i });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', '/store/shop');
  });

  it('renders the correct number of CartCard components when the cart has items', () => {
    // Return a cart with items for this test
    mockUseOutletContext.mockReturnValue({
      cart: [
        { id: 1, name: 'Product 1', price: 10, quantity: 1 },
        { id: 2, name: 'Product 2', price: 20, quantity: 2 },
      ],
      handleCartChange: vi.fn(),
      subtotal: 50,
    });

    render(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>
    );

    expect(screen.getByText('Mock CartCard: Product 1')).toBeInTheDocument();
    expect(screen.getByText('Mock CartCard: Product 2')).toBeInTheDocument();
  });

  it('renders the CartSummary component with the correct subtotal', () => {
    // Return a cart with items for this test
    mockUseOutletContext.mockReturnValue({
      cart: [
        { id: 1, name: 'Product 1', price: 10, quantity: 1 },
        { id: 2, name: 'Product 2', price: 20, quantity: 2 },
      ],
      handleCartChange: vi.fn(),
      subtotal: 50,
    });

    render(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>
    );

    expect(
      screen.getByText('Mock CartSummary: Subtotal is $50')
    ).toBeInTheDocument();
  });

  it('matches snapshot when cart is empty', () => {
    mockUseOutletContext.mockReturnValue({
      cart: [],
      handleCartChange: vi.fn(),
      subtotal: 0,
    });

    const { asFragment } = render(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot when cart has items', () => {
    mockUseOutletContext.mockReturnValue({
      cart: [
        { id: 1, name: 'Product 1', price: 10, quantity: 1 },
        { id: 2, name: 'Product 2', price: 20, quantity: 2 },
      ],
      handleCartChange: vi.fn(),
      subtotal: 50,
    });

    const { asFragment } = render(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
