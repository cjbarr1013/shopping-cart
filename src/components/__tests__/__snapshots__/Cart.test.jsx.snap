// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="_primary_4048ec"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`Cart Component > matches snapshot when cart has items 1`] = `
<DocumentFragment>
  <div
    class="_cartContainer_c7a017"
  >
    <div
      class="_cart_c7a017"
    >
      <div
        class="_cardContainer_c7a017"
      >
        <div>
          Mock CartCard: Product 1
        </div>
        <div>
          Mock CartCard: Product 2
        </div>
      </div>
      <div>
        Mock CartSummary: Subtotal is $50
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cart Component > matches snapshot when cart is empty 1`] = `
<DocumentFragment>
  <div
    class="_emptyCart_c7a017"
  >
    <div
      class="_msg_c7a017"
    >
      <h2>
        Your cart is empty.
      </h2>
      <a
        data-discover="true"
        href="/store/shop"
      >
        Go back to store
      </a>
    </div>
  </div>
</DocumentFragment>
`;
